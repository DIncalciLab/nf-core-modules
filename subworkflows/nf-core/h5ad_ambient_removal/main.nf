include { CELDA_DECONTX               } from '../../../../modules/nf-core/celda/decontx'
include { CELLBENDER_REMOVEBACKGROUND } from '../../../../modules/nf-core/cellbender/removebackground'
include { CELLBENDER_MERGE            } from '../../../../modules/nf-core/cellbender/merge'
include { SOUPX                       } from '../../../../modules/nf-core/soupx'
include { SCVITOOLS_SCAR              } from '../../../../modules/nf-core/scvitools/scar'

workflow H5AD_AMBIENT_REMOVAL {
    take:
    ch_filtered_unfiltered
    tool

    main:
    ch_versions = Channel.empty()

    if (tool == 'none') {
        println "AMBIENT_RNA_REMOVAL: Not performed since 'none' selected."
        ch_h5ad = ch_filtered_unfiltered.map{ meta, filtered, unfiltered -> [meta, filtered] }
    }
    else if (tool == 'decontx') {
        CELDA_DECONTX(ch_filtered_unfiltered)
        ch_h5ad = CELDA_DECONTX.out.h5ad
        ch_versions = ch_versions.mix(CELDA_DECONTX.out.versions)
    }
    else if (tool == 'cellbender') {
        CELLBENDER_REMOVEBACKGROUND(ch_filtered_unfiltered.map{meta, filtered, unfiltered -> [meta, unfiltered]})
        ch_versions = ch_versions.mix(CELLBENDER_REMOVEBACKGROUND.out.versions)

        CELLBENDER_MERGE(ch_filtered_unfiltered.map{ meta, filtered, raw -> [meta.id, meta, filtered, raw] }
            .join(CELLBENDER_REMOVEBACKGROUND.out.h5.map{ meta, h5 -> [meta.id, h5] }, by: 0, failOnMismatch: true)
            .map{ id, meta, filtered, raw, h5 -> [meta, filtered, raw, h5] })
        ch_h5ad = CELLBENDER_MERGE.out.h5ad
        ch_versions = ch_versions.mix(CELLBENDER_MERGE.out.versions)
    }
    else if (tool == 'soupx') {
        SOUPX(ch_filtered_unfiltered)
        ch_h5ad = SOUPX.out.h5ad
        ch_versions = ch_versions.mix(SOUPX.out.versions)
    }
    else if (tool == 'scar') {
        SCVITOOLS_SCAR(ch_filtered_unfiltered)
        ch_h5ad = SCVITOOLS_SCAR.out.h5ad
        ch_versions = SCVITOOLS_SCAR.out.versions
    }
    else {
        error "AMBIENT_REMOVAL: Unexpected tool selected: '${tool}'."
    }

    emit:
    h5ad = ch_h5ad

    versions = ch_versions
}
