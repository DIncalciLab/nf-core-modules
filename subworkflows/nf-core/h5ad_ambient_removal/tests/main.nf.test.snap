{
    "res_none_h5ad": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "/nf-core/test-datasets/scdownstream/samples/SAMN14430799_custom_emptydrops_filter_matrix_5k.h5ad"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-07T08:57:08.637131659"
    },
    "res_scar_h5ad": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.h5ad:md5,222248cba2427bb2dc7f105d285b3536"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-07T09:11:33.310408535"
    },
    "res_soupx_h5ad": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.h5ad:md5,28a794b7b6bf10de5e62489038e0b0e0"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-07T09:12:03.949135183"
    },
    "res_cellbender_h5ad": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.h5ad:md5,77f05ed028b08e0e1f323d52602cdfed"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-07T09:08:32.92735963"
    },
    "res_decontx_h5ad": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.h5ad:md5,ae6bf7c41bcdf1ae3a113fb34a219168"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-07T09:08:57.869483755"
    },
    "res_scar_versions": {
        "content": [
            [
                "versions.yml:md5,9afa250f7c16c9ac819a6b7bd06da540"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-07T09:11:33.710888147"
    },
    "res_cellbender_versions": {
        "content": [
            [
                "versions.yml:md5,2c66ac05eee7711066b1ffc3228db36e",
                "versions.yml:md5,88b58b33730421b2cdcbe88a72b19953"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-07T09:08:32.948181068"
    },
    "res_none_versions": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-07T08:57:08.644321415"
    },
    "res_decontx_versions": {
        "content": [
            [
                "versions.yml:md5,70bf8966ad13fea31521ab54ec5608cb"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-07T09:08:57.926954453"
    },
    "res_soupx_versions": {
        "content": [
            [
                "versions.yml:md5,01dcc1e8fd9f08049ecf9db273284213"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-07T09:12:04.359697329"
    }
}